### PR TITLE
Added test for RewardsReceiver and SharedDepositMinterV2

### PR DESCRIPTION
**Pull request by devlancer412**

### CODE REVIEW

**Comment by chimera-defi** [comment]
looks like CI is failing here
```
  7) SharedDepositMinterV2
       slash:
     Error: Transaction reverted: function selector was not recognized and there's no fallback function
```

**Comment by devlancer412** [comment]
image
Actually I also checked this action. But all test cases are working well on my side. Could you check it on your local?

And seems like it is being reverted at this point
image

But I deployed FeeCalc contract correctly and there aren't any reason to be reverted
image

**Comment by chimera-defi** [comment]
hmm are you sure you uploaded all files? maybe missing local changes?

File: test/v2/core/minter.spec.js [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    const vETH2Addr = "0x898bad2774eb97cf6b94605677f43b41871410b1";
+
+    const Withdrawals = await ethers.getContractFactory("Withdrawals");
+    withdrawals = await Withdrawals.deploy(vETH2Addr, rolloverVirtual);
+    await withdrawals.deployed();
+
+    const numValidators = 1000;
+    const adminFee = 0;
+
+    const FeeCalc = await ethers.getContractFactory("FeeCalc");
```
Comment (chimera-defi) [inline test/v2/core/minter.spec.js]: try to use the zero address for this like the deploy instead. 
the idea was to flesh out and launch a fee calc contract for configuring fees later. but it should default to running internal logic if the passed in address is address(0)

File: test/v2/core/minter.spec.js [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+      value: parseEther("1"),
+    });
+    const afterBalance = await sgEth.balanceOf(wsgEth.address);
+    expect(afterBalance).to.eq(prevBalance.add(parseEther("1")));
+
+    const afterStake = await wsgEth.maxRedeem(deployer.address);
+    expect(afterStake).to.eq(prevStake.add(parseEther("1")));
+  });
+
+  it("depositAndStakeFor", async () => {
```
Comment (chimera-defi) [inline test/v2/core/minter.spec.js]: can we add a test to make sure this works for the direction case? 
i.e. Alice calls the function, but with Bobs address, and bob recvs the tokens in the end

File: test/v2/core/minter.spec.js [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    await expect(minter.connect(alice).withdraw(parseEther("1.1"))).to.be.revertedWith("");
+
+    let prevBalance = await sgEth.balanceOf(alice.address);
+    await minter.connect(alice).withdraw(parseEther("0.5"));
+    let afterBalance = await sgEth.balanceOf(alice.address);
+
+    expect(afterBalance).to.eq(prevBalance.sub(parseEther("0.5")));
+
+    prevBalance = await sgEth.balanceOf(alice.address);
+    await minter.connect(alice).withdrawTo(parseEther("0.5"), alice.address);
```
Comment (chimera-defi) [inline test/v2/core/minter.spec.js]: same as above, can we test this works for directing rewards to others?

File: test/v2/core/minter.spec.js [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    await expect(minter.connect(alice).unstakeAndWithdraw(parseEther("0.5"), alice.address)).to.be.revertedWith("");
+
+    await wsgEth.connect(alice).approve(minter.address, ethers.constants.MaxUint256);
+    let prevBalance = await wsgEth.balanceOf(alice.address);
+    await minter.connect(alice).unstakeAndWithdraw(parseEther("0.5"), alice.address);
+    let afterBalance = await wsgEth.balanceOf(alice.address);
+
+    expect(afterBalance).to.eq(prevBalance.sub(parseEther("0.5")));
+  });
+
```
Comment (chimera-defi) [inline test/v2/core/minter.spec.js:186]: can we add a comment here denoting all the functions below are Admin or Keeper functions? 
And that we are checking that the auth is safe on them?

**Review by chimera-defi — CHANGES_REQUESTED** [review]

**Comment by chimera-defi** [comment]
are we missing anything obvious here in the tests? weird its not showing more like 100% cov

```
File                                          |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |

 contracts/v2/core/                           |    66.67 |    46.81 |    68.89 |    61.54 |                |
  RewardsReceiver.sol                         |    83.33 |       50 |       60 |    77.78 |          42,48 |
  Rollover.sol                                |        0 |        0 |        0 |        0 |    28,33,34,37 |
  SgETH.sol                                   |      100 |      100 |      100 |      100 |                |
  SharedDepositMinterV2.sol                   |    73.81 |       50 |    79.17 |    65.71 |... 260,262,266 |
```

**Comment by devlancer412** [comment]
Actually I'm not sure what should I test on Rollover contract. it's just have one internal function.
About Minter contract, I couldn't test some functions because it's connected to StakingContract(seems like it is external contract).
About RewardsReceiver, I tested all. seems like another functions are ownable functions such as transferOwner, renounceOwnaship

**Comment by chimera-defi** [comment]
ah ok, can you add tests for the owner functions, check to see they work only for the owner?  @devlancer412 
other than that, just comments on the code review

File: test/v2/core/minter.spec.js [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    const NOR_ROLE = await minter.NOR();
+    await expect(minter.connect(alice).setWithdrawalCredential("0x")).to.be.revertedWith(
+      `AccessControl: account ${alice.address.toLowerCase()} is missing role ${NOR_ROLE}`,
+    );
+
+    await minter.setWithdrawalCredential("0x");
+  });
+
+  it("slash", async () => {
+    const GOV_ROLE = await minter.GOV();
```
Comment (chimera-defi) [inline test/v2/core/minter.spec.js]: example: 
- user deposits
- rewards generated 
- reward sync called 
- slash called
- user redeems 
ensure user does not get back >1

File: test/v2/core/minter.spec.js (hunk 4 above)
Comment (devlancer412) [reply test/v2/core/minter.spec.js:150]: What's the admin or keeper functions? and I'm not sure what are you saying with auth. you mean ROLES  such as ADMIN, GOV and NOR?

**Comment by devlancer412** [comment]
@chimera-defi  Please check last updates

**Comment by chimera-defi** [comment]
nice, can we fix the build plz? @devlancer412

**Comment by devlancer412** [comment]
@chimera-defi  merged another test pr to here

**Review by chimera-defi — APPROVED** [review]

File: contracts/lib/xERC4626.sol [hunk 6]
```diff
@@ -12,95 +12,95 @@ import {IxERC4626} from "../interfaces/IxERC4626.sol";
...
 
-    if (timestamp < rewardsCycleEnd) revert SyncError();
+        storedTotalAssets = storedTotalAssets_ + lastRewardAmount_; // SSTORE
 
-    uint256 storedTotalAssets_ = storedTotalAssets;
-    uint256 nextRewards = asset.balanceOf(address(this)) - storedTotalAssets_ - lastRewardAmount_;
+        uint32 end = ((timestamp + rewardsCycleLength) / rewardsCycleLength) * rewardsCycleLength;
 
-    storedTotalAssets = storedTotalAssets_ + lastRewardAmount_; // SSTORE
+        if (end - timestamp < rewardsCycleLength / 20) {
```
Comment (chimera-defi) [inline contracts/lib/xERC4626.sol:95]: whats with the changes here? can we reset this file to head? 
ill add any auto lint stuff later

**Review by chimera-defi — APPROVED** [review]

File: contracts/lib/xERC4626.sol (hunk 6 above)
Comment (devlancer412) [reply contracts/lib/xERC4626.sol:95]: don't have any changes on this file. all changes are from prettier such as removing spaces

**Pull request merged**